File: app/src/main/java/com/example/pam_project/di/ApplicationContainerLocator.java
package com.example.pam_project.di;

import android.content.Context;

import androidx.annotation.VisibleForTesting;

public class ApplicationContainerLocator {

    private static ApplicationContainer applicationContainer;

    private ApplicationContainerLocator() {}

    public static ApplicationContainer locateComponent(final Context context) {
        if (applicationContainer == null) {
            setComponent(new ProductionApplicationContainer(context));
        }

        return applicationContainer;
    }

    @VisibleForTesting
    public static void setComponent(final ApplicationContainer applicationContainer) {
        ApplicationContainerLocator.applicationContainer = applicationContainer;
    }
}

File: app/src/main/java/com/example/pam_project/di/ApplicationContainerLocator.kt
package com.example.pam_project.di

import android.content.Context
import androidx.annotation.VisibleForTesting

object ApplicationContainerLocator {
    private var applicationContainer: ApplicationContainer? = null
    fun locateComponent(context: Context): ApplicationContainer? {
        if (applicationContainer == null) {
            setComponent(ProductionApplicationContainer(context))
        }
        return applicationContainer
    }

    @VisibleForTesting
    fun setComponent(applicationContainer: ApplicationContainer?) {
        ApplicationContainerLocator.applicationContainer = applicationContainer
    }
}